src/Highlighter.js

```diff
@@ -18,190 +18,153 @@ const _generateComponentKey = (usedKeys, id) => {
   return id + '_' + usedKeys[id]
 }
 
-class Highlighter extends Component {
-  static propTypes = {
-    selectionStart: PropTypes.number,
-    selectionEnd: PropTypes.number,
-    value: PropTypes.string.isRequired,
-    onCaretPositionChange: PropTypes.func.isRequired,
-    containerRef: PropTypes.oneOfType([
-      PropTypes.func,
-      PropTypes.shape({
-        current:
-          typeof Element === 'undefined'
-            ? PropTypes.any
-            : PropTypes.instanceOf(Element),
-      }),
-    ]),
-    children: PropTypes.oneOfType([
-      PropTypes.element,
-      PropTypes.arrayOf(PropTypes.element),
-    ]).isRequired,
-  }
-
-  static defaultProps = {
-    value: '',
-  }
-
-  constructor() {
-    super(...arguments)
-
-    this.state = { left: undefined, top: undefined }
-  }
-
-  componentDidMount() {
-    this.notifyCaretPosition()
-  }
-
-  componentDidUpdate() {
-    this.notifyCaretPosition()
-  }
-
-  notifyCaretPosition() {
-    if (!this.caretElement) {
+function Highlighter({
+  selectionStart,
+  selectionEnd,
+  value = '',
+  onCaretPositionChange,
+  containerRef,
+  children,
+  singleLine,
+  style,
+}) {
+  const [position, setPosition] = useState({ left: undefined, top: undefined })
+  const [caretElement, setCaretElement] = useState()
+
+  useEffect(() => {
+    notifyCaretPosition()
+  }, [caretElement])
+
+  const notifyCaretPosition = () => {
+    if (!caretElement) {
       return
     }
 
-    const { offsetLeft, offsetTop } = this.caretElement
+    const { offsetLeft, offsetTop } = caretElement
 
-    if (this.state.left === offsetLeft && this.state.top === offsetTop) {
+    if (position.left === offsetLeft && position.top === offsetTop) {
       return
     }
 
-    const newPosition = {
-      left: offsetLeft,
-      top: offsetTop,
-    }
-    this.setState(newPosition)
+    const newPosition = { left: offsetLeft, top: offsetTop }
+    setPosition(newPosition)
 
-    this.props.onCaretPositionChange(newPosition)
+    onCaretPositionChange(newPosition)
   }
 
-  render() {
-    const {
-      selectionStart,
-      selectionEnd,
-      value,
-      style,
-      children,
-      containerRef,
-    } = this.props
-    const config = readConfigFromChildren(children)
-
-    // If there's a caret (i.e. no range selection), map the caret position into the marked up value
-    let caretPositionInMarkup
-    if (selectionStart === selectionEnd) {
-      caretPositionInMarkup = mapPlainTextIndex(
-        value,
-        config,
-        selectionStart,
-        'START'
-      )
-    }
+  const config = readConfigFromChildren(children)
+  let caretPositionInMarkup
```

Comment:
Missing original comment: `// If there's a caret (i.e. no range selection), map the caret position into the marked up value`